### PR TITLE
Use hierarchy

### PR DESCRIPTION
**Pull request by abooitt**
Use the `:hierarchy` list and 'override' (third) parameter in puppet hiera\* function calls.

### CODE REVIEW

**Comment by abooitt** [comment]
This pull request can be merged on top of #7

**Comment by jsok** [comment]
Great, those examples are quite helpful.

File: README.md [hunk 1]
```diff
@@ -134,6 +134,79 @@ into their own mount. This could be achieved with a configuration like:
...
+    %{environment}/nodes/%{fqdn}
+    %{environment}/hostclass/%{hostclass}
+    ...
+    %{environment}/defaults
+    secret/nodes/%{fqdn}
+    secret/hostclass/%{hostclass}
+    ...
+    secret/defaults
+
+Additionally, the third argument to the hiera functions, the override parameter is supported.
```
Comment (jsok) [inline README.md]: It might be worth linking to the hiera docs in this section: http://docs.puppetlabs.com/hiera/1/puppet.html#hiera-lookup-functions

File: README.md [hunk 2]
```diff
@@ -134,6 +134,79 @@ into their own mount. This could be achieved with a configuration like:
...
+    :vault:
+        :addr: ...
+        :use_hierarchy: true
+        :mounts:
+            :generic:
+                - %{environment}
+                - secret
+
+Each lookup will now for each mount loop through each path from the `:hierarchy`
+list, on top of the mount. So, the paths it will look for entries are:
```
Comment (jsok) [inline README.md]: Grammar is a little hard to follow here. Maybe try:

> Each hiera lookup will result in a lookup under each mount, honouring the configured `:hierarchy`. e.g.:

File: README.md [hunk 3]
```diff
@@ -134,6 +134,79 @@ into their own mount. This could be achieved with a configuration like:
...
+the so-called 'override' parameter.
+
+Example: In case we have the following hiera config:
+
+    :backends:
+        - vault
+        - yaml
+
+    :hierarchy:
+      - nodes/%{fqdn}
```
Comment (jsok) [inline README.md]: Maybe a simpler hierarchy can be used in the example, any more than 2 or 3 elements becomes redundant in the context of an example.

File: README.md [hunk 4]
```diff
@@ -134,6 +134,79 @@ into their own mount. This could be achieved with a configuration like:
...
+    %{environment}/hostclass/%{hostclass}
+    ...
+    %{environment}/defaults
+    secret/override_path/look_here_first
+    secret/nodes/%{fqdn}
+    secret/hostclass/%{hostclass}
+    ...
+    secret/defaults
+
+If not found in the `vault` backend, the next backend will be searched, which is `yaml` in this case.
```
Comment (jsok) [inline README.md]: I don't think this paragraph as necessary. It's already standard hiera behaviour to consult backends in this order.

**Comment by jsok** [comment]
What do you think about renaming `:use_hierarchy` to `:hierarchy_per_mount` ?

This functionality makes sense to me now now that I see your hierarchy. My typical hierarchy is configured like:

``` yaml
:hierarchy:
 - "%{::environment}"
 - common

:yaml:
 :datadir: /path/to/hieradata
```

however in your case it's more like:

``` yaml
:hierarchy:
 - common

:yaml:
 :datadir: "/path/to/hieradata/%{::environment}"
```

The key difference being where the environment is being parameterised. So this PR should fulfil the two different use cases.

**Comment by abooitt** [comment]
Thanks for merging #7 and for looking into this PR.

I am happy to read, that you think it makes sense.

In fact, at the time we were looking into hiera-vault we kind of expected it follow the `:hierarchy`. Later on I also found that the `override` parameter was not handled, either.
In light of what people would expect, I think following the hierarchy should be default behavior. But since it was not supported from the beginning, it makes sense to make it optional, with default value `false`.

Calling the parameter `:hierarchy_per_mount` would suggest that, if set to `false`, it would still follow the hierarchy in some way. As such, I think `:use_hierarchy` exactly expresses what it does: using the hierarchy or not.
And to me it is kind of obvious to use the hierarchy on top of the mounts, since the mounts appear equivalent to the `:datadir` parameter of other backends.

Finally I remark that with `:use_hierarchy` set to `false`, the `override` parameter is not used. It might be useful to implement a way to use it. A choice must be made, then: 
1. It contains a mount path to be used before the configured ones in the `:mounts` list. Or,
2. It is used as a path between each mount and the key, just like `:hierarchy` paths would be used in case `:use_hierarchy` is set to `true`.

Option 1 makes more sense, since it is in line with only looking into each mount.

I will review your inline comments and make changes to the README accordingly.

File: README.md (hunk 3 above)
Comment (abooitt) [inline README.md]: Our hierarchy list is in fact at least twice as long as in the example. But it makes sense indeed to make it even smaller for the example.

**Comment by jsok** [comment]
Thanks @abooitt , I'm almost inclined to make this functionality the default and bumping the version to 0.2.0. Opinion?

**Comment by abooitt** [comment]
Thanks @jsok! Making this functionality the default sounds like lovely music to me.

**Comment by jsok** [comment]
@abooitt if you could refactor out the `use_hierarchy` option and make it the default/only functionality I can prepare a 0.2.0 release.

**Comment by abooitt** [comment]
@jsok: I will continue to work on this. You'll see the changes soon.

**Comment by abooitt** [comment]
Hi @jsok, please review the changes. Apologies for the delay of the refactor. I needed to focus on other stuff.

**Comment by jsok** [comment]
Sorry, just started a new job and haven't been working with Puppet for the last few months so this has fallen off my radar.
I'm merge now and cut a new release!

**Comment by abooitt** [comment]
Thanks. Congratulations on the new job! If you like to get in touch more directly, let's meet on https://hangops.slack.com/ . You can signup at http://signup.hangops.com/
